secio/src/codec/hmac_compat/openssl_impl.rs

```diff
@@ -0,0 +1,64 @@
+use openssl::{
+    hash::MessageDigest,
+    memcmp,
+    pkey::{PKey, Private},
+    sign::Signer,
+};
+
+use crate::Digest;
+
+/// Hash-based Message Authentication Code (HMAC).
+pub struct Hmac {
+    digest: MessageDigest,
+    key: PKey<Private>,
+}
+
+impl Hmac {
+    /// Returns the size of the hash in bytes.
+    #[inline]
+    pub fn num_bytes(&self) -> usize {
+        self.digest.size()
+    }
+
+    /// Builds a `Hmac` from an algorithm and key.
+    pub fn from_key(algorithm: Digest, key: &[u8]) -> Self {
+        let digest = match algorithm {
+            Digest::Sha256 => MessageDigest::sha256(),
+            Digest::Sha512 => MessageDigest::sha512(),
+        };
+
+        let key = PKey::hmac(key).unwrap();
+        Hmac { digest, key }
+    }
+
+    /// Signs the data.
+    pub fn sign(&mut self, crypted_data: &[u8]) -> Vec<u8> {
+        Signer::new(self.digest, &self.key)
+            .unwrap()
+            .sign_oneshot_to_vec(crypted_data)
+            .unwrap()
```

Comment:
I don't understand why `unwrap` can be used here, how do we ensure that the `crypted_data` length is checked?

Reply:
>  how do we ensure that the crypted_data length is checked?

I don't understand your question, HMRC has no requirement for the length of the input data, so why check its length, it is a process similar to the hash, the output data length varies depending on the algorithm

> I don't understand why unwrap can be used here

Here to ensure that the input meets the requirements of the OpenSSL premise, I can not find the error way, which can be changed to `expect`

Reply:
I found a description of the hmac input length: https://crypto.stackexchange.com/questions/10034/hmac-sha1-input-size

In our usage scenario, this problem does not exist, its input data is the shared key after ecdh, this key is different because of different algorithms, but all belong to small data

Reply:
I may have misunderstood the purpose of this fn, can you indicate where these two fn `Hmac#sign` and `Hmac#verify` are used in this PR? I don't see where they are called specifically other than in unit tests.

Reply:
> I may have misunderstood the purpose of this fn, can you indicate where these two fn Hmac#sign and Hmac#verify are used in this PR? 

In the [real code](https://github.com/nervosnetwork/tentacle/blob/master/secio/src/handshake/procedure.rs#L271), the context is used and there is no place to use sign and verify. They exist only to verify that the different implementations are consistent

Reply:
understood, it's better to add a `#[test]` to these fn if they are only used in unit test, public fn may lead to misuse by users, for example, it will panic when use the `verify` fn with two different length slice.

Reply:
I added a todo to hide hmac in the next break change version, which is currently exposed to the outside world, and modified the logic of verify to verify the length first and then give it to the eq function to avoid panic